notebooks/autodiff_cookbook.ipynb

```diff
@@ -0,0 +1,1252 @@
+    "        # Use vmap to do a matrix-Jacobian product.\n",
+    "        # Here, the matrix is the Euclidean basis, so we get all\n",
+    "        # entries in the Jacobian at once. \n",
+    "        J, = vmap(vjp_fun, in_axes=0)(np.eye(len(y)))\n",
+    "        return J\n",
+    "    return jacfun\n",
+    "\n",
+    "assert np.allclose(builtin_jacrev(f)(W), our_jacrev(f)(W)), 'Incorrect reverse-mode Jacobian results!'"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 20,
+   "metadata": {
+    "colab": {},
+    "colab_type": "code",
+    "id": "Qd9gVZ5t6vvP"
+   },
+   "outputs": [],
+   "source": [
+    "from jax import jacfwd as builtin_jacfwd\n",
+    "\n",
+    "def our_jacfwd(f):\n",
+    "    def jacfun(x):\n",
+    "        _jvp = lambda s: jvp(f, (x,), (s,))[1]\n",
+    "        Jt =vmap(_jvp, in_axes=1)(np.eye(len(x)))\n",
+    "        return np.transpose(Jt)\n",
+    "    return jacfun\n",
+    "\n",
+    "assert np.allclose(builtin_jacfwd(f)(W), our_jacfwd(f)(W)), 'Incorrect forward-mode Jacobian results!'"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Interestingly, [Autograd](https://github.com/hips/autograd) couldn't do this. Our [implementation of reverse-mode `jacobian` in Autograd](https://github.com/HIPS/autograd/blob/96a03f44da43cd7044c61ac945c483955deba957/autograd/differential_operators.py#L60) had to pull back one vector at a time with an outer-loop `map`. Pushing one vector at a time through the computattion is much less efficient than batching it all together with `vmap`."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Another thing that Autograd couldn't do is `jit`. Interestingly, no matter how much Python dynamism you use in your function to be differentiated, we could always use `jit` on the linear part of the computation. For example:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 21,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "(array(3.1415927, dtype=float32),)\n"
+     ]
+    }
+   ],
+   "source": [
+    "def f(x):\n",
+    "    try:\n",
+    "        if x < 3:\n",
+    "            return 2 * x ** 3\n",
+    "        else:\n",
+    "            raise ValueError\n",
+    "    except ValueError:\n",
+    "        return np.pi * x\n",
+    "\n",
+    "y, f_vjp = vjp(f, 4.)\n",
+    "print(f_vjp(1.))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Complex numbers and differentiation"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "JAX is great at complex numbers and differentiation. To support both holomorphic and non-holomorphic differentiation, JAX follows [Autograd's convention](https://github.com/HIPS/autograd/blob/master/docs/tutorial.md#complex-numbers) for encoding complex derivatives.\n",
```

Comment:
You might link to "holomorphic function" on Wikipedia the first time you mention it, if, as I do, you need a refresher in complex analysis.